### PR TITLE
Save for later show more

### PR DESCRIPTION
**Pull request by NathanielBennett**
This paginates the saved for later page and includes changes in the branch saved-for-later-show-on-fronts. Here, we show four items per page, so the controller methods change to deal with this. This means that the previous javascript functionality for deleting items is removed. The delete all button now has a confirm state, which, when pressed for a second time ( guess what ) deletes all stories. All buttons now make a round server trip: 

Note: This follow on for this PR: ps://github.com/guardian/frontend/pull/9391 so on opening includes those changes. Styling to follow in a further PR.

save-for-later-list

### CODE REVIEW

**Comment by robertberry** [comment]
Might wanna get design to look at the positioning of that delete button. cc @sammorrisdesign

**Comment by NathanielBennett** [comment]
@robertberry the styling's going to come in a separate PR

**Comment by robertberry** [comment]
You should add correct image widths at various breakpoints here: https://github.com/guardian/frontend/blob/master/common/app/layout/WidthsByBreakpoint.scala

It's used for outputting the `srcset` and `sizes` attribute of images. Without it, you might not get high res images, or the browser might end up downloading larger images than it needs.

File: identity/app/controllers/SaveContentController.scala [hunk 1]
```diff
@@ -65,7 +61,7 @@ class SaveContentController @Inject() ( api: IdApiClient,
             if (!prefs.contains(shortUrl)) {
               val articleId = idRequest.articleId match {
                 case Some(id) => id
-                case _ =>  new URL(returnUrl).getPath.drop(1)
+                case _ => new URL(returnUrl).getPath.drop(1)
```
Comment (robertberry) [inline identity/app/controllers/SaveContentController.scala]: Never use `URL`, always `URI`. `URL` is evil. It does weird unexpected side effects like DNS lookups!

File: identity/app/controllers/SaveContentController.scala [hunk 2]
```diff
@@ -79,32 +75,31 @@ class SaveContentController @Inject() ( api: IdApiClient,
...
-        fillFormWithApiDataAndGetResult(idRequest, form, savedArticles)
-       case Left(errors) =>
+        fillFormWithApiDataForPageAndGetResult(idRequest, savedArticles, pageNum)
+      case Left(errors) =>
         val formWithErrors = errors.foldLeft(savedArticlesForm) {
           case (formWithErrors, Error(message, decription, _, context)) =>
             formWithErrors.withError(context.getOrElse(""), message)
         }
-        Future.successful(NoCache(Ok(views.html.profile.savedForLater(page, formWithErrors, List.empty, formActionUrl(idUrlBuilder, idRequest, "/saved-for-later")))))
+        val pageData = pageDataBuilder.apply(List.empty, emptyArticles, idRequest, pageNum)
```
Comment (robertberry) [inline identity/app/controllers/SaveContentController.scala]: Why are you calling `apply` here? Can you not just call `pageDataBuilder` directly, like a function?

File: identity/app/controllers/SaveContentController.scala [hunk 3]
```diff
@@ -113,67 +108,78 @@ class SaveContentController @Inject() ( api: IdApiClient,
...
+            val updatedArticlesViow: Option[Future[Result]] = data.deleteArticle.map {
+              shortUrlOfDeletedArticle =>
+                val updatedArticles = savedArticles.removeArticle(shortUrlOfDeletedArticle)
+                val updatedResult = api.updateSavedArticles(request.user.auth, updatedArticles).flatMap {
+                  case Right(updatedArticles) =>
+                    fillFormWithApiDataForPageAndGetResult(idRequest, updatedArticles, pageNum)
+
+                  case Left(errors) =>
+                    val formWithApiErrors = buildFormFromErrors(errors)
+                    val pageData = pageDataBuilder.apply(List.empty, emptyArticles, idRequest, pageNum)
```
Comment (robertberry) [inline identity/app/controllers/SaveContentController.scala]: Ditto to above

File: identity/app/controllers/SaveContentController.scala [hunk 4]
```diff
@@ -113,67 +108,78 @@ class SaveContentController @Inject() ( api: IdApiClient,
...
 
-              updatedArticlesViow.getOrElse {
-                val formWithError = form.withError("Error", "There was a problem with your request")
-                Future.successful(
-                  NoCache(Ok(views.html.profile.savedForLater(page, form, List.empty, formActionUrl(idUrlBuilder, idRequest, "/saved-for-later"))))
-                )
-              }
+            updatedArticlesViow.getOrElse {
+              val formWithError = form.withError("Error", "There was a problem with your request")
+              val pageData = pageDataBuilder.apply(List.empty, emptyArticles, idRequest, pageNum)
```
Comment (robertberry) [inline identity/app/controllers/SaveContentController.scala]: Ditto

File: identity/app/controllers/SaveContentController.scala [hunk 5]
```diff
@@ -113,67 +108,78 @@ class SaveContentController @Inject() ( api: IdApiClient,
...
-        getResponse(LiveContentApi.item(article.id, Edition.defaultEdition).showFields("webTitle,webUrl,trailText,shortUrl").showElements("all")).map(_.content.map(Content(_)))
     }
 
-    Future.sequence(savedApiContentItems).map { savedForLaterSeq =>
-      val contentList = savedForLaterSeq.toList.collect {
-        case Some(content) => content
+  private def fillFormWithApiDataForPageAndGetResult(idRequest: IdentityRequest, updatedArticles: SavedArticles, pageNum: Int)
+                                                    (implicit request: RequestHeader): Future[Result] = {
+
+    //Deal with case where one item on last page has` been deleted
```
Comment (robertberry) [inline identity/app/controllers/SaveContentController.scala]: Stray `

File: identity/app/implicits/Articles.scala [hunk 6]
```diff
@@ -15,8 +15,16 @@ trait Articles {
 
   implicit class RichSavedArticles(savedArticles: SavedArticles) {
     val fmt = ISODateTimeFormat.dateTimeNoMillis()
+    val itemsPerPage = 5
 
+    val pages = savedArticles.articles match {
+      case Nil => List.empty
```
Comment (robertberry) [inline identity/app/implicits/Articles.scala]: What's the point of this branch?

File: identity/app/implicits/Articles.scala [hunk 7]
```diff
@@ -41,6 +49,16 @@ trait Articles {
...
     }
+
+    //Deal with just having removed the only item on the last page
+    def getPage(page: Int): List[SavedArticle] = pages match {
+      case Nil => List.empty
+      case _ => page >= pages.length match {
+        case true => pages.last
+        case _ =>pages(page - 1)
+      }
+
```
Comment (robertberry) [inline identity/app/implicits/Articles.scala]: This is a bit over-complicated.

```scala
def getPage(page: Int) = 
    pages.lift(Math.min(page, pages.length - 1)).getOrElse(Nil)
```

File: identity/app/views/fragments/profile/saveForLater/savedArticleImage.scala.html [hunk 8]
```diff
@@ -0,0 +1,16 @@
+@import model.Content
+@import views.support.ImgSrc
+@import layout.ContentWidths.BodyMedia
+
+@(article: Content)(implicit request: RequestHeader)
+
+<div class="saved-content__img media__img">
```
Comment (robertberry) [inline identity/app/views/fragments/profile/saveForLater/savedArticleImage.scala.html]: What's the reason for this separate template?

**Comment by robertberry** [comment]
Looks good to me otherwise :+1:

File: identity/app/implicits/Articles.scala (hunk 7 above)
Comment (robertberry) [inline identity/app/implicits/Articles.scala]: Actually even simpler:

```scala
def getPage(page: Int) = pages.lift(page) orElse pages.lastOption getOrElse Nil
```

**Pull request merged**